server/src/main/java/org/opensearch/action/search/SearchRequestOperationsListener.java

```diff
@@ -32,6 +33,35 @@ void onRequestStart(SearchRequestContext searchRequestContext) {}
 
     void onRequestEnd(SearchPhaseContext context, SearchRequestContext searchRequestContext) {}
 
+    public void setEnabled(boolean enabled) {
+        if (enabled) {
+            register();
+        } else {
+            deregister();
+        }
+    }
```

Comment:
Rather than doing this `register`/`deregister` dance, which is difficult to do from a thread-safety standpoint (without triggering a `ConcurrentModificationException`), could we add a volatile `enabled` property with a package-private getter? Every listener can be registered with the listener manager once (on construction).

When we add listeners to a specific instance of `TransportSearchAction`, the `SearchRequestListenerManager` can check `isEnabled` and only add the enabled listeners. (Optionally, the `isEnabled` method could take the current `SearchRequest` and decide based on some combination of the setting and the request.)

Reply:
Thanks for bringing this up! I also had some hesitation between "using isEnable and add all listeners to the manager" or "only adding those listeners when they are explicitly enabled". Yes, it's a good point that it would be better to not do the "register" workflow considering the multithreading factor. Let me refactor this part of code.